File: Easy/1-TwoSum.cc

#include <vector>
#include <unordered_map>
#include <iostream>
using namespace std;

//Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.
//You may assume that each input would have exactly one solution, and you may not use the same element twice.
//You can return the answer in any order. 

// More details: https://leetcode.com/problems/two-sum/description/



/*
    Solution:

    Time complexity: O(n).
    We traverse the list containing n elements only once. Each lookup in the table costs only O(1) time.

    Space complexity: O(n).
    The extra space required depends on the number of items stored in the hash table, which stores at most nnn elements.
*/

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> hash;
        // Loop through each number in the vector.
        for (int i = 0; i < nums.size(); i++) {
            // Calculate the complement of the current number.
            // If nums[i] is a part of the solution, the other part is target - nums[i].
            int complement = target - nums[i]; 
            if (hash.find(complement) != hash.end()) {
                // Return the indices of the two numbers that add up to the target.
                // hash[complement] is the index of the complement,
                // and i is the current index.
                return {hash[complement], i};
            }
            // If the complement is not found, store the current number and its index in the map.
            // This is for future reference when we encounter the complement.
            hash[nums[i]] = i;
        }
        // If no solution is found, return an empty vector. This should never happen.
        return {};
    }
};


/*
    Submission   98.46% faster than other C++ submissions (27/06/2024):


class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> hash;
        for (int i = 0; i < nums.size(); i++) {
            int complement = target - nums[i]; 
            if (hash.find(complement) != hash.end()) 
                return {hash[complement], i};
            hash[nums[i]] = i;
        }
        return {};
    }
};


*/


//Test
int main() {
    Solution solution;
    vector<int> nums = {2, 7, 11, 15}; // 2 + 7 = 9 => Indices: 0 and 1
    int target = 9;
    vector<int> result = solution.twoSum(nums, target);
    if (!result.empty()) {
        cout << "Indices: " << result[0] << " and " << result[1] << endl;
    } else {
        cout << "No two sum solution found." << endl;
    }
    return 0;
}

File: Python/Easy/1-TwoSum.py
# Description: https://leetcode.com/problems/two-sum/description/

"""
    Approach: Hash Map

    Use a dictionary to store each number and its index as we iterate.
    For each element, calculate its complement (target - current number).
    Check if the complement already exists in the dictionary.
    If found, return the complement's index and current index.
    Otherwise, store the current number and index in the dictionary.
    This achieves single-pass lookup with O(1) average time per operation.

    Time complexity: O(n) - Single pass through the array.
    Space complexity: O(n) - Dictionary stores at most n elements.
"""

from typing import List, Dict

# Solution
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        num_map: Dict[int, int] = {}
        for i, num in enumerate(nums):
            complement = target - num
            if complement in num_map:
                return [num_map[complement], i]
            num_map[num] = i
        return []

# Test
if __name__ == '__main__':
    solution = Solution()
    nums = [2, 7, 11, 15]
    target = 9
    result = solution.twoSum(nums, target)
    print(f"Input: nums = {nums}, target = {target}")
    print(f"Output: {result}")
